### PR TITLE
adopt torch.testing.assert_close

### LINKED ISSUE #1029: Be an early adopter of `torch.testing.assert_close`
## 🚀 Feature

Be an early adopter of `torch.testing.assert_close`.

## Motivation

We (PyTorch) recently added [`torch.testing.assert_close`](https://pytorch.org/docs/1.9.0/testing.html) that will ship with the upcoming `torch==1.9.0` release in a prototype state. We are looking for early adopters to get feedback for missing features or UI / UX improvements. 

## Pitch

`kornia` uses the undocumented `torch.testing.assert_allclose` (note the additional `all`) throughout the test suite. It performs similar but less checks, and gives less expressive error messages in case anything goes wrong. Plus, we plan on deprecating / removing it in the future in favor of `torch.testing.assert_close` so you would have to switch eventually.

I've run the test suite against a PyTorch nightly build and replaced all occurrences of `torch.testing.assert_allclose` with `torch.testing.assert_close`, and > 300 errors came up. Most of them are related to mismatching strides of tensors. It would be a good datapoint for us if these are false positives for you.

If you have any comments on the design, please feel free to give us feedback in pytorch/pytorch#56544.

### PR DESCRIPTION
**Pull request by pmeier**
Closes #1029. `torch.testing.assert_close` will only be shipped with `torch==1.9`. Thus, this PR should not be merged before.

All files except for `test/utils.py` are 1-to-1 ports and probably do not need special review, but I let you be the judge of that.

### CODE REVIEW

File: test/utils.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        actual: torch.Tensor,
+        expected: torch.Tensor,
+        *,
+        rtol: Optional[float] = None,
+        atol: Optional[float] = None,
+        **kwargs: Any,
+    ) -> None:
+        if rtol is None and atol is None:
+            with contextlib.suppress(Exception):
+                rtol, atol = _get_default_tolerance(actual, expected)
```
Comment (pmeier) [inline test/utils.py:23]: `torch.testing.assert_close` uses stricter tolerances as `torch.testing.assert_allclose`. There are < 10 tests that fail with the new default tolerances. Thus we use the old ones here.

**Comment by shijianjian** [comment]
Just one quick question, would it break the backward compatibility in the CI?

**Comment by pmeier** [comment]
> Just one quick question, would it break the backward compatibility in the CI?

No, it won't. I've backported it using `torch.testing.assert_allclose` in case `torch.testing.assert_close` is not available.

**Comment by edgarriba** [comment]
thanks @pmeier ! this is a great contribution from the pytorch team :)

**Comment by Borda** [comment]
> No, it won't. I've backported it using `torch.testing.assert_allclose` in case `torch.testing.assert_close` is not available.

so is `torch.testing.assert_close` missing in torch<1.9?

**Comment by pmeier** [comment]
> so is `torch.testing.assert_close` missing in torch<1.9?

Yes. It will be shipped the first time with `torch==1.9.0`.

**Comment by Borda** [comment]
> > so is `torch.testing.assert_close` missing in torch<1.9?
> 
> Yes. It will be shipped the first time with `torch==1.9.0`.

co can you run all these updated tests in past PT 1.7 for example?

**Comment by pmeier** [comment]
The CI already does that, doesn't it?

- [`torch==1.6.0`](https://github.com/kornia/kornia/pull/1031/checks?check_run_id=2658680745)
- [`torch==1.7.0`](https://github.com/kornia/kornia/pull/1031/checks?check_run_id=2658680794)
- [`torch==1.7.1`](https://github.com/kornia/kornia/pull/1031/checks?check_run_id=2658680820)
- [`torch==1.8.0`](https://github.com/kornia/kornia/pull/1031/checks?check_run_id=2658680836)

**Comment by Borda** [comment]
> The CI already does that, doesn't it?

that is something that makes me suspicious, how it can use a future function which is not existing in that codebase? :confused:

**Comment by edgarriba** [comment]
@pmeier @Borda we disabled mult-version in the CI to lighten the load

**Comment by pmeier** [comment]
@Borda See https://github.com/kornia/kornia/pull/1031#issue-651550603

> `torch.testing.assert_close` will only be shipped with `torch==1.9`. Thus, this PR should not be merged before.

I've just prepared the PR so this is ready when `kornia` officially support `torch==1.9`. I can confirm that it runs fine with the pre-release, but we need for CI to confirm that.

**Comment by edgarriba** [comment]
I guess that the point of @Borda is that will this PR won't have backward compatibility

**Comment by pmeier** [comment]
Why wouldn't it? CI is green right now for all the older versions.

**Comment by edgarriba** [comment]
@pmeier you are right. Was confused by a moment with @Borda comments. The PR runs on a lighter version of the full version matrix. See, https://github.com/kornia/kornia/blob/master/.github/workflows/tests_cpu.yml#L50

@Borda what was your point here ? 🤔

**Comment by Borda** [comment]
> I've just prepared the PR so this is ready when `kornia` officially supports `torch==1.9`.

But we cannot use it till Kornia will set 1.9 as min version...

**Comment by pmeier** [comment]
@Borda before we continue this cycle, please have a look at the file `test/utils.py` that will be added by this PR.

> > I've just prepared the PR so this is ready when `kornia` officially supports `torch==1.9`.
> 
> But we cannot use it till Kornia will set 1.9 as min version...

We can use it for all CI runs that run on `torch>=1.9`. Currently, there is no such run, but there will be in the near future. 

> > @Borda what was your point here ? thinking
> 
> As mentioned, the CI for PR will pass but not on mater, and you can't test any older PT versions anymore...

Of course it will pass on `master`. I did not touch the minimum requirement of `torch` and the CI is green.

> > The PR runs on a lighter version of the full version matrix.
> 
> So how is it possible that your tests are using function that shall not exist in the particular PT version?

The tests aren't doing that. As I have explained in https://github.com/kornia/kornia/pull/1031#issuecomment-847605666 I've added a minimal backport of `torch.testing.assert_close` in case it is not available. So all our current tests are using this backport.

**Comment by Borda** [comment]
> @Borda before we continue this cycle, please have a look at the file `test/utils.py` that will be added by this PR.

yeah, I missed this part, and in fact, I was just backing about this point, sorry for the confusion... :rabbit:

**Comment by edgarriba** [comment]
@pmeier @Borda actually forgot that we run against nightlies in daily basis and everything seems fine 
https://github.com/kornia/kornia/runs/2724828690?check_suite_focus=true

**Comment by pmeier** [comment]
@edgarriba @Borda please let me know when the CI includes `torch>=1.9` so I can update the PR.

**Comment by edgarriba** [comment]
@pmeier https://github.com/kornia/kornia/pull/1120 all yours !

**Comment by edgarriba** [comment]
@pmeier I think you need to rebase so that can trigger 1.9 tests

**Comment by pmeier** [comment]
Yes, but I didn't have time yet to do so. I'll request another review when the PR is ready.

**Comment by pmeier** [comment]
One thing that is different from the old `assert_allclose` to the new `assert_close` is the default tolerance. We run some diagnostics and set the new tolerances a little stricter than the old ones. The current adoption reuses the old tolerances. If we were to use the new ones, these tests would fail.

```
test/augmentation/test_augmentation_mix.py::TestRandomCutMix::test_random_mixup_beta0
test/augmentation/test_random_generator.py::TestRandomPerspectiveGen::test_random_gen
test/augmentation/test_random_generator.py::TestRandomPerspectiveGen::test_same_on_batch
test/augmentation/test_random_generator.py::TestRandomRotationGen::test_random_gen
test/augmentation/test_random_generator.py::TestRandomRotationGen::test_random_gen
test/augmentation/test_random_generator.py::TestRandomRotationGen::test_same_on_batch
test/color/test_hls.py::TestRgbToHls::test_unit
test/color/test_hls.py::TestHlsToRgb::test_unit
test/color/test_hsv.py::TestRgbToHsv::test_unit
test/enhance/test_adjust.py::TestAdjustSaturation::test_saturation_one
test/enhance/test_adjust.py::TestAdjustSaturation::test_saturation_one_batch
test/enhance/test_adjust.py::TestAdjustHue::test_hue_one
test/enhance/test_adjust.py::TestAdjustHue::test_hue_one_batch
test/feature/test_matching.py::TestMatchSMNN::test_matching1
test/filters/test_sobel.py::TestSobel::test_magnitude
test/filters/test_sobel.py::TestSobel::test_magnitude
test/geometry/transform/test_imgwarp.py::TestWarpPerspective::test_crop_center_resize
test/geometry/transform/crop/test_crop3d.py::TestCropAndResize3D::test_crop
test/geometry/transform/crop/test_crop3d.py::TestCropAndResize3D::test_crop_batch
test/integration/test_conversions.py::TestAngleOfRotations::test_matrix_angle
test/utils/test_metrics.py::TestConfusionMatrix::test_three_classes_normalized
```

@edgarriba could you have a look if there is anything special to these tests? If not can I just crank up the tolerances in there and use the new defaults everywhere else?

**Review by edgarriba — APPROVED** [review]

File: kornia/testing/__init__.py [hunk 2]
```diff
@@ -128,3 +130,44 @@ def _get_precision_by_name(
         return tol_val
 
     return tol_val_default
+
+
+try:
```
Comment (edgarriba) [inline kornia/testing/__init__.py:135]: @pmeier probably we should add here a note or todo saying that if any time we do not support <1.9 - we should directly use `torch.testing.assert_close` ?

**Comment by edgarriba** [comment]
> One thing that is different from the old `assert_allclose` to the new `assert_close` is the default tolerance. We run some diagnostics and set the new tolerances a little stricter than the old ones. The current adoption reuses the old tolerances. If we were to use the new ones, these tests would fail.
> 
> ```
> test/augmentation/test_augmentation_mix.py::TestRandomCutMix::test_random_mixup_beta0
> test/augmentation/test_random_generator.py::TestRandomPerspectiveGen::test_random_gen
> test/augmentation/test_random_generator.py::TestRandomPerspectiveGen::test_same_on_batch
> test/augmentation/test_random_generator.py::TestRandomRotationGen::test_random_gen
> test/augmentation/test_random_generator.py::TestRandomRotationGen::test_random_gen
> test/augmentation/test_random_generator.py::TestRandomRotationGen::test_same_on_batch
> test/color/test_hls.py::TestRgbToHls::test_unit
> test/color/test_hls.py::TestHlsToRgb::test_unit
> test/color/test_hsv.py::TestRgbToHsv::test_unit
> test/enhance/test_adjust.py::TestAdjustSaturation::test_saturation_one
> test/enhance/test_adjust.py::TestAdjustSaturation::test_saturation_one_batch
> test/enhance/test_adjust.py::TestAdjustHue::test_hue_one
> test/enhance/test_adjust.py::TestAdjustHue::test_hue_one_batch
> test/feature/test_matching.py::TestMatchSMNN::test_matching1
> test/filters/test_sobel.py::TestSobel::test_magnitude
> test/filters/test_sobel.py::TestSobel::test_magnitude
> test/geometry/transform/test_imgwarp.py::TestWarpPerspective::test_crop_center_resize
> test/geometry/transform/crop/test_crop3d.py::TestCropAndResize3D::test_crop
> test/geometry/transform/crop/test_crop3d.py::TestCropAndResize3D::test_crop_batch
> test/integration/test_conversions.py::TestAngleOfRotations::test_matrix_angle
> test/utils/test_metrics.py::TestConfusionMatrix::test_three_classes_normalized
> ```
> 
> @edgarriba could you have a look if there is anything special to these tests? If not can I just crank up the tolerances in there and use the new defaults everywhere else?

@pmeier probably it's worth opening an issue and mention that we should take a look at this tests and their implementations.

**Pull request merged**